packages/alfa-aria/src/feature.ts

```diff
@@ -657,49 +657,29 @@ const Features: Features = {
     tfoot: html(() => Option.of(Role.of("rowgroup"))),
 
     th: html(
-      (element) => {
-        const table = element
+      (element) =>
+        element
           .ancestors()
           .filter(isElement)
-          .find(hasName("table"));
-
-        // If the <th> is not in a <table>, it doesn't really have a role…
-        if (table.isNone()) {
-          return None;
-        }
-
-        const tableModel = Table.from(table.get());
-
-        // If the <th> is within a <table> with errors, it doesn't really have a
-        // role.
-        if (tableModel.isErr()) {
-          return None;
-        }
-
-        const cell = Iterable.find(tableModel.get().cells, (cell) =>
-          cell.element.equals(element)
-        );
-
-        // If the current element is not a cell in the table, something weird
-        // happened and it doesn't have a role.
-        if (cell.isNone()) {
-          return None;
-        }
-
-        // This is not fully correct. If the header has no variant, its role
-        // should be computed as a <td>
-        // https://www.w3.org/TR/html-aam-1.0/#html-element-role-mappings
-        return cell.get().scope.map((scope) => {
-          switch (scope) {
-            case Scope.Column:
-            case Scope.ColumnGroup:
-              return Role.of("columnheader");
-            case Scope.Row:
-            case Scope.RowGroup:
-              return Role.of("rowheader");
-          }
-        });
-      },
+          .find(hasName("table"))
+          .map(Table.from)
+          .flatMap((table) =>
+            table.cells.filter(Cell.isHeader).find(Cell.hasElement(element))
+          )
+          .map((cell) => {
+            switch (cell.scope) {
+              case "column":
+              case "column-group":
+                return Role.of("columnheader");
+
+              case "row":
+              case "row-group":
+                return Role.of("rowheader");
+
+              default:
+                return Role.of("cell");
```

Comment:
This actually depends on the role of the table element (its `gridcell` if the table is a `grid`).
```suggestion
for (const [role] of Role.from(table)) {
              if (role.isSome()) {
                switch (role.get().name) {
                  case "table":
                    return Role.of("cell");
                  case "grid":
                    return Role.of("gridcell");
                }
              }
            }
```

Reply:
Good point, I'll get that cleared up!